fee-hook/src/FeeHook.sol

```diff
@@ -0,0 +1,208 @@
+        if (uncollectedFees[poolId] >= MIN_WITHDRAW) {
+            // 1. Collect fees.
+            bytes memory actions = abi.encodePacked(uint8(Actions.DECREASE_LIQUIDITY), uint8(Actions.TAKE_PAIR));
+            bytes[] memory params = new bytes[](2);
+            uint256 tokenId = TOKEN_ID[poolId];
+            params[0] = abi.encode(tokenId, 0, 0, 0, bytes(""));
+            params[1] = abi.encode(key.currency0, key.currency1, address(this));
+            posm.modifyLiquiditiesWithoutUnlock(
+                actions, params
+            );
+
+            // 2. Transfer fees to this contract on the destination chain.
+            if (block.chainid != destinationChainId) {
+                bytes memory payload = abi.encode();
+                // FIXME tune this value, could push it lower
+                uint128 messageFee = 3_000_000_000_000_000;
+                uint128 GAS_LIMIT = 500000;
+                uint128 VALUE = uint128(address(this).balance);
+                bytes memory options = OptionsBuilder.newOptions().addExecutorLzReceiveOption(GAS_LIMIT, VALUE - messageFee);
+                // Note that external payable functions cannot be called directly
+                // within a contract, so we have to instantiate a new instance
+                // of the FeeHook contract to call the sendCrossChain function.
+                FeeHook(payable(address(this))).sendCrossChain{value: VALUE}(
+                    payload,
+                    options,
+                    VALUE,
+                    address(this)
+                );
+            }
+            uncollectedFees[poolId] = 0;
+
+        }
+        return (BaseHook.afterSwap.selector, 0);
+    }
+
+    function _beforeAddLiquidity(
+        address,
+        PoolKey calldata key,
+        IPoolManager.ModifyLiquidityParams calldata,
+        bytes calldata
+    ) internal override returns (bytes4) {
+        PoolId poolId = key.toId();
+        if (TOKEN_ID[poolId] == 0) {
+            TOKEN_ID[poolId] = posm.nextTokenId() - 1;
+            poolIds.push(PoolId.unwrap(poolId));
+        }
+        return BaseHook.beforeAddLiquidity.selector;
+    }
+
+    // Allows the owner to transfer an LP position to another address
+    // in case any manual intervention is needed.
+    function transferPosition(
+        address to,
+        bytes32 poolId) external onlyOwner
+    {
+        PoolId id = PoolId.wrap(poolId);
+        uint256 tokenId = TOKEN_ID[id];
+        require(tokenId != 0, "Token ID not found");
+        transferredPoolIds.push(poolId);
+        PositionManager(payable(address(posm))).safeTransferFrom(address(this), to, tokenId, "");
+    }
+
+
+    function withdrawFees() external {
+        if (block.chainid != destinationChainId) {
+            revert("Not on destination chain. Cannot withdraw fees");
+        }
+        uint256 totalSupply = IERC20(vMooneyAddress).totalSupply();
```

Comment:
Would this get the total supply on that chain? Or across chains?

Reply:
Reason being because someone might have a high proportion on Base, but their "total" proportion considering mainnet would be much lower

Reply:
need to come up with a way to address this. i think we can do it with layerzero.